torch/_inductor/triton_heuristics.py

```diff
@@ -165,10 +166,56 @@ def precompile(self, warm_cache_only_with_cc=None):
         with self.lock:
             if self.launchers:
                 return
-            self.launchers = [
-                self._precompile_config(c, warm_cache_only_with_cc)
-                for c in self.configs
-            ]
+            self.launchers = []
+            compiled_binaries = []
+            for c in self.configs:
+                compiled_binary, launcher = self._precompile_config(
+                    c, warm_cache_only_with_cc
+                )
+                self.launchers.append(launcher)
+                compiled_binaries.append(compiled_binary)
+
+            if (
+                config.dynamic_scale_rblock
+                and self.heuristic_type == HeuristicType.REDUCTION
+                and self.size_hints is not None
+            ):
+                for triton_config, compiled_binary in zip(
+                    self.configs, compiled_binaries
+                ):
+                    assert len(self.size_hints) == 2
+                    xblock = triton_config.kwargs["XBLOCK"]
+                    rblock = triton_config.kwargs["RBLOCK"]
+                    total_block = (self.size_hints[0] + xblock - 1) // xblock
+                    nreg = getattr(compiled_binary, "n_regs", None)
+                    if nreg is None:
+                        continue
+
+                    # each SM of A100 has 65536 32-bit registers. To maximize
+                    # the theoretical occupancy, we need run 2048 threads on each
+                    # SM. So each thread should use no more than 65536 / 2048
+                    # = 32 registers. In cases where occupancy matters, and each
+                    # thread uses too many registers, reduce RBLOCK to reduce
+                    # the register usage.
+                    # For kernel https://gist.github.com/shunting314/e4cccc031fe30d378b9b23c08c238cbd
+                    # from PLBartForCausalLM, latency improve from
+                    # 7.795ms to 4.883ms.
+                    if nreg <= 32:
+                        continue
+                    max_blocks_per_sm = 2048 // (32 * triton_config.num_warps)
+                    # TODO: this is hardcoded for A100, we should revise for H100
+                    # for optimal perf.
+                    if total_block <= max_blocks_per_sm * 108:
+                        # <100% occupancy is fine
+                        continue
+                    # make sure rblock is not too small
+                    if rblock <= 64:
+                        continue
```

Comment:
In this case, wouldn't we want to decrease `XBLOCK`?

Reply:
yea, we could in order to reduce register usages for each thread. But in this case, we also gonna have more threads and blocks. Occupancy is improved but we may still need the same number of waves to run all the thread blocks.

Also, in most cases I checked, our XBLOCK is usually 1 :) We can not decrease further. I think it's due to this heuristic: https://github.com/pytorch/pytorch/blob/d4990ad5a14fa2352ce3fff4f469cf499167c1a3/torch/_inductor/triton_heuristics.py#L903